pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -43,43 +42,81 @@ import (
 	"istio.io/istio/pkg/spiffe"
 	"istio.io/istio/pkg/util/sets"
 	"istio.io/istio/pkg/workloadapi"
+	"istio.io/istio/pkg/workloadapi/security"
 )
 
 // AmbientIndex maintains an index of ambient WorkloadInfo objects by various keys.
 // These are intentionally pre-computed based on events such that lookups are efficient.
 type AmbientIndex struct {
 	mu sync.RWMutex
-	// byService indexes by Service (virtual) *IP address*. A given Service may have multiple IPs, thus
+	// byService indexes by network/Service (virtual) *IP address*. A given Service may have multiple IPs, thus
 	// multiple entries in the map. A given IP can have many workloads associated.
-	byService map[string][]*model.WorkloadInfo
-	// byPod indexes by Pod IP address.
+	byService map[networkAddress][]*model.WorkloadInfo
+	// byPod indexes by network/podIP address.
 	byPod map[string]*model.WorkloadInfo
+	// services are indexed by the network/clusterIP
+	services map[networkAddress]*model.ServiceInfo
 
 	// Map of ServiceAccount -> IP
 	// TODO: currently, this is derived from pods. To be agnostic to the implementation,
 	// we should actually be looking at Gateway.status.addresses.
 	// This may be an external address (possibly even a DNS name we need to resolve), an arbitrary IP,
 	// or a reference to a service.
 	// If its a reference to a Service then we can find the underlying pods in that service, as an optimization.
-	waypoints map[model.WaypointScope]sets.String
+	waypoints map[model.WaypointScope]*workloadapi.GatewayAddress
 
 	// serviceVipIndex maintains an index of VIP -> Service
 	serviceVipIndex *kclient.Index[string, *v1.Service]
 }
 
 // Lookup finds a given IP address.
-func (a *AmbientIndex) Lookup(ip string) []*model.WorkloadInfo {
+func (a *AmbientIndex) Lookup(key string) []*model.AddressInfo {
 	a.mu.RLock()
 	defer a.mu.RUnlock()
 	// First look at pod...
-	if p, f := a.byPod[ip]; f {
-		return []*model.WorkloadInfo{p}
+	if p, f := a.byPod[key]; f {
+		addr := &workloadapi.Address{
+			Type: &workloadapi.Address_Workload{
+				Workload: p.Workload,
+			},
+		}
+		return []*model.AddressInfo{{Address: addr}}
 	}
+
 	// Fallback to service. Note: these IP ranges should be non-overlapping
-	return a.byService[ip]
+	// cannot distunagish between Service lookup and Workloads for a Service lookup
+	// since the same newtork/vip key is used.  Both Service and Workloads are
+	// returned in this case
```

Comment:
see the comment above: "Fallback to service. Note: these IP ranges should be non-overlapping"

If this were to happen then I think we have a logic error and did not validate in the control plane properly before indexing

Reply:
Ah I think this comment is just confusing. Perhaps a clearer version of the comment: "When we look up a service we will also want to return its workloads"

Reply:
updated comment